### ISSUE TITLE
Short term planning

### ISSUE THREAD

**Issue by davidmorgan** [dart-lang/macros#84]
@jakemac53 I thought maybe we can try tracking in-flight PRs and what to work on today/tomorrow in an issue.

## Out for review

#71 not blocking anything, convenience extension methods

## Next steps / WIP

High priority

- Push on #93
- Address TODO in macro_implementation.dart, "assumes augmentations are for the macro application target", pass the augmentation target explicitly in `Augmentation`.
- Get started on TODOs "helper for augmenting methods" in json_codable.dart, i.e. "builders"
   - Support for getting `QualifiedName` for `member` in #87 may be helpful
- Add support for macros reporting errors
   - Add support for e2e errors tests.
      - TODOs related to validation in json_codable.dart e.g. `_checkNamedType`
- Kick off prototyping around macro metadata 

Medium priority

- Push on #94, next step is to add rerunning macros, e2e benchmarking
- Address TODO about `supertypeOf` in `type_system.dart` added in #89 
- Bring CFE impl in line with analyzer impl i.e. equivalent of changes to analyzer impl in #68 
   - Phases handling: equivalent of https://dart-review.googlesource.com/c/sdk/+/394740
   - Then ensure golden tests are in sync
- Add e2e tests using "apply using CFE" option in `_macro_tool`
- Model evolution: add `since` tags and defaults to fields, add minor version numbers, update generation to use them
- Address TODO in `macro/lib/macro.dart`: inject a macro's metadata from the bootstrap script rather than writing macro name and phases directly in the macro.
- (see #85 #86) Build out query language + data model: expand `Query`, `Interface`, `Member` to union types.
- Switch e2e test from sh script to dart

Low priority

- Iterate on `JsonBufferBuilder` e.g. #83
- TODO in `macro_client.dart`: split to multiple implementations for TCP/IP, stdin/out, etc. (Isolate mode?)
- TODO in `macro_client.dart`, various others: handle macro lifecycles so it's possible to edit a macro and have it rebuild.
   - Make an analysis server binary with the host code injected so it's possible to see it work live in an IDE.
- TODOs in `generate_dart_model.dart` "add the description" and "add description(s)".

**Comment by jakemac53** [comment]
This works, but I am not sure how we want to manage not stepping on each others toes?

- we could create an issue for each and assign ourselves
- we could just edit the comment above and add our name to an item if we start working on it

Maybe also change each thing to a checkbox so we can mark them as completed

**Comment by davidmorgan** [comment]
Let's try editing the comment above, I started with "WIP @davidmorgan" against the "add support for macros reporting errors" :)

I was thinking of deleting done stuff but if it's useful (it probably is useful to see for a while at least) we could also move it to a "done" section as an alternative to checkboxes.

**Comment by jakemac53** [comment]
> I was thinking of deleting done stuff but if it's useful (it probably is useful to see for a while at least) we could also move it to a "done" section as an alternative to checkboxes.

Or just strike through them 🤷

**Comment by davidmorgan** [comment]
Most items are blocked on the two pending PRs; I could work on top of them but I don't want to build up a big pile of unreviewed changes, so I'll switch to doing cleanup / docs / other things for now.

**Comment by simolus3** [comment]
If this is not being worked on already, I would like to contribute the query-related part of

> Build out query language + data model: expand `Query`, `Interface`, `Member` to union types.

As a first step (if I understand things correctly), I'd split the `Query` class into a union that includes the current `Query`,  a query to only resolve the static type part (useful when macros only care about types but not about the class structure) and a query that includes multiple other queries to send them in a single request.

**Comment by davidmorgan** [comment]
Thanks @simolus3 :) a good start!

I created #86 to discuss, please note that Jake's away for a few days so there will be some delay getting his input, then I'll be away for the week of the 14th which slow things down again, sorry about that.
